### PR TITLE
fix issue #1056

### LINKED ISSUE #1056: The program may need a check step
### Description
The program may need a check step.

### Steps to reproduce
1. Convert a png file into a pdf file forcibly
2. Use pdf to image
3. Choose the pdf file which converted before

### Expected behaviour
Maybe output that the file is not a pdf file.

### Actual behaviour
Action succeeded and needs a password.

### Would you like to work on the issue?
No, I am just a user.

test

### PR DESCRIPTION
**Pull request by Jay-Ernest**
# Description

Please include a summary of the change and which issue is fixed. List any dependencies that are required for this change.
If there are any UI change, please include the screenshots also.

Fixes #1056 

## Type of change
Just put an x in the [] which are valid.
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes.
- [x] `./gradlew assembleDebug assembleRelease`
- [x] `./gradlew checkstyle`

# Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings

### CODE REVIEW

**Comment by Jay-Ernest** [comment]
# Description
## problem
The program doesn't check whether the pdf file is  valid.
https://github.com/Swati4star/Images-to-PDF/issues/1056  

## how to solve
I added several methods to check the head of the file whether it is '%PDF-1.4'  

## the other issue may produce
Maybe it will still lead to OOM while using the app.  

## where I test the app
Using the virtual machine: Pixel 4 API 30

File: app/src/main/java/swati4star/createpdf/adapter/MergeFilesAdapter.java [hunk 1]
```diff
@@ -46,12 +48,61 @@ public ViewMergeFilesHolder onCreateViewHolder(@NonNull ViewGroup parent, int vi
...
+            boolean isEncrypted = mPDFUtils.isPDFEncrypted(mFilePaths.get(position));
+            holder.mFileName.setText(FileUtils.getFileName(mFilePaths.get(position)));
+            holder.mEncryptionImage.setVisibility(isEncrypted ? View.VISIBLE : View.INVISIBLE);
+        }
+    }
+
+    // check whether the file is a real pdf file or not
+    public static boolean isPdfFile(String filePath) {
+        String value = getFileHeader(filePath);
+        return value.startsWith("255044462D312E");
```
Comment (codegsaini) [inline app/src/main/java/swati4star/createpdf/adapter/MergeFilesAdapter.java:63]: Can you explain this please!

**Comment by Jay-Ernest** [comment]
Normally，a pdf file is a file with a head with “%PDF1.4”. But if I just changing the file suffix，the file head will not change into “%PDF1.4”，it is still in the encoding mode before，so while the program reading the file，it will recognize it as pdf，but infact，the file is a png or other encoding mode. So the program will lead to IOException，and pdfEncryption = true，makes the program think the pdf is encrypted. But I cannot fix this thoroughly，maybe I add the check in some wrong place？



发自我的iPhone


------------------ Original ------------------
From: Gaurav Saini ***@***.***&gt;
Date: Sat, Apr 23, 2022 3:39 PM
To: Swati4star/Images-to-PDF ***@***.***&gt;
Cc: Jay-Ernest ***@***.***&gt;, Author ***@***.***&gt;
Subject: Re: [Swati4star/Images-to-PDF] fix issue #1056 (PR #1057)





 
@codegsaini commented on this pull request.
 
 
In app/src/main/java/swati4star/createpdf/adapter/MergeFilesAdapter.java:
 &gt; @@ -46,12 +48,61 @@ public ViewMergeFilesHolder ***@***.*** ViewGroup parent, int vi        @Override      public void ***@***.*** ViewMergeFilesHolder holder, int position) { -        boolean isEncrypted = mPDFUtils.isPDFEncrypted(mFilePaths.get(position)); -        holder.mFileName.setText(FileUtils.getFileName(mFilePaths.get(position))); -        holder.mEncryptionImage.setVisibility(isEncrypted ? View.VISIBLE : View.INVISIBLE); +        boolean isPdfFile = isPdfFile(mFilePaths.get(position)); +        if (isPdfFile) { +            boolean isEncrypted = mPDFUtils.isPDFEncrypted(mFilePaths.get(position)); +            holder.mFileName.setText(FileUtils.getFileName(mFilePaths.get(position))); +            holder.mEncryptionImage.setVisibility(isEncrypted ? View.VISIBLE : View.INVISIBLE); +        } +    } + +    // check whether the file is a real pdf file or not +    public static boolean isPdfFile(String filePath) { +        String value = getFileHeader(filePath); +        return value.startsWith("255044462D312E");  
Can you explain this please!

File: app/src/main/java/swati4star/createpdf/adapter/MergeFilesAdapter.java (hunk 1 above)
Comment (Jay-Ernest) [reply app/src/main/java/swati4star/createpdf/adapter/MergeFilesAdapter.java:63]: maybe the comment should be changed into 'check the file head making sure that the file is encoding in pdf mode'

Comment (Jay-Ernest) [reply app/src/main/java/swati4star/createpdf/adapter/MergeFilesAdapter.java:63]: and if it is not in pdf encoding mode, the program will not set the file visible to the user

Comment (Jay-Ernest) [reply app/src/main/java/swati4star/createpdf/adapter/MergeFilesAdapter.java:63]: the pdf head “%PDF1.4” in hex is "255044462D312E"

**Comment by codegsaini** [comment]
> Normally，a pdf file is a file with a head with “%PDF1.4”. But if I just changing the file suffix，the file head will not change into “%PDF1.4”，it is still in the encoding mode before.

Ok that seems good validation technique but I think that we should not check **%PDF1.4** (with specific version number) rather we should only compare **%PDF** hex to determine whether it's a PDF or not.

[According to wikipedia[link]](https://en.wikipedia.org/wiki/PDF/A#:~:text=ISO%2019005%20%E2%80%93%20Document,%5B5%5D) **1.4** here is representing  the ISO-standardized version of pdf format which is earlier version and recent in **2.0** So we should omit version from our validation.

[For reference, you can check out this Stackoverflow answer[link].](https://stackoverflow.com/a/13297022)

**Comment by codegsaini** [comment]
> But I cannot fix this thoroughly，maybe I add the check in some wrong place ?

@Jay-Ernest Dive deep into the source code and check out where app can accept PDF as input and place validation at those places.

Why hurry, take your time. :)

**Comment by Jay-Ernest** [comment]
> > Normally，a pdf file is a file with a head with “%PDF1.4”. But if I just changing the file suffix，the file head will not change into “%PDF1.4”，it is still in the encoding mode before.
> 
> Ok @Jay-Ernest, that seems good validation technique but I think that we should not check **%PDF1.4** (with specific version number) rather we should only compare **%PDF** hex to determine whether it's a PDF or not.
> 
> [According to wikipedia[link]](https://en.wikipedia.org/wiki/PDF/A#:~:text=ISO%2019005%20%E2%80%93%20Document,%5B5%5D) **1.4** here is representing the ISO-standardized version of pdf format which is earlier version and recent in **2.0** So we should omit version from our validation.
> 
> [For reference, you can check out this Stackoverflow answer[link].](https://stackoverflow.com/a/13297022)

Yes, I agree with you. So I changed my code in #1062 to check the header %PDF. Thank you for the reference.

**Comment by Jay-Ernest** [comment]
> > But I cannot fix this thoroughly，maybe I add the check in some wrong place ?
> 
> @Jay-Ernest Dive deep into the source code and check out where app can accept PDF as input and place validation at those places.
> 
> Why hurry, take your time. :)

Emm, it is my first time to participate in Android, I find it difficult for me to deal with the problem, like how to pop out a window to alert the user to select the correct file.

**Comment by codegsaini** [comment]
> how to pop out a window to alert the user to select the correct file.

Or can we just show a message that selected pdf is not valid?

**Comment by Jay-Ernest** [comment]
> > how to pop out a window to alert the user to select the correct file.
> 
> Or can we just show a message that selected pdf is not valid? We can show alert popup by using `AlertDialog` class. Check [stackoverflow answer for reference.](https://stackoverflow.com/a/2115770)

Thank you. I will try this.

**Pull request closed without merging**